File: chatgpt-plugin/mcp-server/src/knowledge.ts
import { readFile } from "node:fs/promises";
import { join } from "node:path";

export const documentPaths = [
  "AGENTS.md",
  "docs/agent-rules.md",
  "docs/question-routing.md",
  "docs/metric-definitions.md",
  "docs/data-dictionary.md",
  "docs/agent-reporting.md",
  "api/openapi.yaml",
  "docs/REPO_AUDIT.md",
  "docs/MCP_TOOLS.md",
  "docs/REPORTING_METRICS.md",
];

export type KnowledgeDocument = { id: string; title: string; url: string };

export async function loadDocuments(root: string): Promise<Map<string, string>> {
  const documents = new Map<string, string>();
  for (const path of documentPaths) {
    try {
      documents.set(path, await readFile(join(root, path), "utf8"));
    } catch {
      // Optional generated docs are absent before the first build.
    }
  }
  return documents;
}

export function searchDocuments(
  documents: Map<string, string>,
  query: string,
): { results: KnowledgeDocument[] } {
  const terms = query.toLowerCase().match(/[a-z0-9_]+/g) ?? [];
  const ranked = [...documents.entries()]
    .map(([id, text]) => ({
      score: terms.reduce((total, term) => total + text.toLowerCase().split(term).length - 1, 0),
      id,
    }))
    .filter((item) => item.score > 0)
    .sort((a, b) => b.score - a.score || a.id.localeCompare(b.id))
    .slice(0, 8);
  return { results: ranked.map(({ id }) => documentSummary(id)) };
}

export function fetchDocument(
  documents: Map<string, string>,
  id: string,
): KnowledgeDocument & { text: string; metadata: { source: string } } {
  const text = documents.get(id);
  if (!text) throw new Error("Unknown knowledge document");
  return { ...documentSummary(id), text, metadata: { source: "data-reporting-kit" } };
}

function documentSummary(id: string): KnowledgeDocument {
  return {
    id,
    title: id.split("/").at(-1) ?? id,
    url: `https://github.com/lightningtransport/data-reporting-kit/blob/main/${id}`,
  };
}

File: chatgpt-plugin/mcp-server/src/knowledge.py
"""Local, revisioned documentation retrieval for the ChatGPT MCP plugin."""
from __future__ import annotations

import re
from collections.abc import Mapping
from typing import Any
from urllib.parse import quote

_TOKEN = re.compile(r"[a-z0-9_]+")


class KnowledgeBase:
    def __init__(self, documents: Mapping[str, str], canonical_base_url: str) -> None:
        self.documents = dict(documents)
        self.canonical_base_url = canonical_base_url.rstrip("/")

    def search(self, query: str, limit: int = 8) -> dict[str, list[dict[str, str]]]:
        terms = set(_TOKEN.findall(query.lower()))
        matches: list[tuple[int, str]] = []
        for document_id, text in self.documents.items():
            score = sum(text.lower().count(term) for term in terms)
            if score:
                matches.append((score, document_id))
        matches.sort(key=lambda item: (-item[0], item[1]))
        return {"results": [self._result(document_id) for _, document_id in matches[:limit]]}

    def fetch(self, document_id: str) -> dict[str, Any]:
        if document_id not in self.documents:
            raise ValueError("Unknown knowledge document")
        result = self._result(document_id)
        return {**result, "text": self.documents[document_id], "metadata": {"source": "data-reporting-kit"}}

    def _result(self, document_id: str) -> dict[str, str]:
        return {
            "id": document_id,
            "title": document_id.rsplit("/", 1)[-1],
            "url": f"{self.canonical_base_url}/{quote(document_id)}",
        }
